src/Lecture2.hs

```diff
@@ -288,4 +392,18 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding expression = buildExp (sum literals) variables
+  where (literals, variables) = go [] [] expression
+        go :: [Int] -> [String] -> Expr -> ([Int], [String])
+        go lits vars (Var x)     = (lits, x:vars)
```

Comment:
For performance reasons, you can already store `Int` instead of `[Int]` here and replace starting `[]` with zero. So you don't need to store all the constants, you can add them immediately and store only a single number on each step instead of the list of numbers.